test/system/helpers.network.bash

```diff
@@ -0,0 +1,297 @@
+    local protocol=${3:-tcp}
+
+    local port
+    for port in $(shuf -i ${range}); do
+        if port_is_free $port $address $protocol; then
+            echo $port
+            return
+        fi
+    done
+
+    die "Could not find open port in range $range"
+}
+
+# random_free_port_range() - Get range of unbound ports with pseudorandom start
+# $1:	Size of range (i.e. number of ports)
+# $2:	Optional binding address, any IPv4 address by default
+# $3:	Optional protocol, tcp or udp
+function random_free_port_range() {
+    local size=${1?Usage: random_free_port_range SIZE [ADDRESS [tcp|udp]]}
+    local address=${2:-0.0.0.0}
+    local protocol=${3:-tcp}
+
+    local maxtries=10
+    while [[ $maxtries -gt 0 ]]; do
+        local firstport=$(random_free_port)
+        local lastport=
+        for i in $(seq 1 $((size - 1))); do
+            lastport=$((firstport + i))
+            if ! port_is_free $lastport $address $protocol; then
+                echo "# port $lastport is in use; trying another." >&3
+                lastport=
+                break
+            fi
+        done
+        if [[ -n "$lastport" ]]; then
+            echo "$firstport-$lastport"
+            return
+        fi
+
+        maxtries=$((maxtries - 1))
+    done
+
+    die "Could not find free port range with size $size"
+}
+
+# port_is_bound() - Check if TCP or UDP port is bound for a given address
+# $1:	Port number
+# $2:	Optional protocol, or optional IPv4 or IPv6 address, default: tcp
+# $3:	Optional IPv4 or IPv6 address, or optional protocol, default: any
+function port_is_bound() {
+    local port=${1?Usage: port_is_bound PORT [tcp|udp] [ADDRESS]}
+
+    if   [ "${2}" = "tcp" ] || [ "${2}" = "udp" ]; then
+        local address="${3}"
+        local proto="${2}"
+    elif [ "${3}" = "tcp" ] || [ "${3}" = "udp" ]; then
+        local address="${2}"
+        local proto="${3}"
+    else
+        local address="${2}"	# Might be empty
+        local proto="tcp"
+    fi
+
+    port=$(printf %04X ${port})
+    case "${address}" in
+    *":"*)
+        grep -e "^[^:]*: $(ipv6_to_procfs "${address}"):${port} .*" \
+             -e "^[^:]*: $(ipv6_to_procfs "::0"):${port} .*"        \
+             -q "/proc/net/${proto}6"
+        ;;
+    *"."*)
+        grep -e "^[^:]*: $(ipv4_to_procfs "${address}"):${port}"    \
+             -e "^[^:]*: $(ipv4_to_procfs "0.0.0.0"):${port}"       \
+             -q "/proc/net/${proto}"
+        ;;
+    *)
+        # No address: check both IPv4 and IPv6, for any bound address
+        grep "^[^:]*: [^:]*:${port} .*" -q "/proc/net/${proto}6" || \
+        grep "^[^:]*: [^:]*:${port} .*" -q "/proc/net/${proto}"
+        ;;
+    esac
+}
+
+# port_is_free() - Check if TCP or UDP port is free to bind for a given address
+# $1:	Port number
+# $2:	Optional protocol, or optional IPv4 or IPv6 address, default: tcp
+# $3:	Optional IPv4 or IPv6 address, or optional protocol, default: any
+function port_is_free() {
+    ! port_is_bound ${@}
```

Comment:
Hint for the future: the only way to use dollar-at is in double quotes: `"$@"`. Any other use is dangerous: it does not properly expand spaces. Not a concern here, but it's important to get in the habit.

Reply:
No, that was really my intention. It's not about "properly expand[ing] spaces", it's about [field splitting](https://pubs.opengroup.org/onlinepubs/9699919799.2013edition/utilities/V3_chap02.html#tag_18_05_02) (there's no globbing here, I refer to the POSIX specification because it's clearer, but Bash doesn't differ here). I want to make sure arguments are split in case somebody happens to pass `"5000 tcp"` instead of bothering them just for the sake of ShellCheck's SC2068 or the like.

"Hint for the future" and "get in the habit" coupled with non-technical arguments sounds rather condescending, by the way. It would help me interacting if you refrained from doing that. I don't see any added value.